File: library/src/main/java/com/yy/codex/uikit/CALayerPainter.java
package com.yy.codex.uikit;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.graphics.PorterDuffXfermode;
import android.graphics.RectF;
import android.support.annotation.NonNull;

/**
 * Created by adi on 17/1/17.
 */

public class CALayerPainter {

    private static Paint sPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public static void drawLayerTree(CALayer layer, Canvas canvas){
        Bitmap srcBitmap = createBitmapWithLayerTree(layer);
        Bitmap maskBitmap = createBitmapWithMaskLayer(layer);
        Bitmap resultBitmap = createBitmapThatApplyMask(layer, srcBitmap, maskBitmap);
        sPaint.reset();
        canvas.drawBitmap(resultBitmap, 0, 0, sPaint);
    }

    public static void drawCurrentLayer(CALayer layer, Canvas canvas){
        if (layer.isHidden()){
            return;
        }
        if (layer.getCornerRadius() > 0){
            drawRoundRect(canvas, layer);
            if (layer.getBitmap() != null){
                drawRoundRectBitmap(canvas, layer);
            }
            if (layer.getBorderWidth() > 0){
                drawRoundRectBorder(canvas, layer);
            }
        }
        else {
            drawRect(canvas, layer);
            if (layer.getBitmap() != null){
                drawRectBitmap(canvas, layer);
            }
            if (layer.getBorderWidth() > 0){
                drawRectBorder(canvas, layer);
            }
        }

        if (layer.getMask() != null){
            // @TODO
        }
    }

    /* support method */

    private static void drawRoundRect(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);
        float borderWidth = (float) layer.getBorderWidth() * scaledDensity;
        float cornerRadius = (float) layer.getCornerRadius() * scaledDensity;
        float halfBorderW = borderWidth / 2.0f;

        sPaint.reset();
        sPaint.setAntiAlias(true);
        sPaint.setColor(layer.getBackgroundColor().toInt());
        RectF rectFCopyed = frame.shrinkToRectF(halfBorderW, origin);
        if (layer.getShadowRadius() > 0) {
            float shadowRadius = (float) layer.getShadowRadius() * scaledDensity;
            float shadowX = (float) layer.getShadowX() * scaledDensity;
            float shadowY = (float) layer.getShadowY() * scaledDensity;
            sPaint.setShadowLayer(shadowRadius, shadowX, shadowY, layer.getShadowColor().toInt());
            rectFCopyed = new RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY);
        }
        canvas.drawRoundRect(rectFCopyed, cornerRadius, cornerRadius, sPaint);
    }

    private static void drawRoundRectBitmap(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        Bitmap bitmap = layer.getBitmap();
        CALayer.BitmapGravity bitmapGravity = layer.getBitmapGravity();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);
        float cornerRadius = (float) layer.getCornerRadius() * scaledDensity;
        UIColor bitmapColor = layer.getBitmapColor();

        CGRect maskFrame = new CGRect(origin.x, origin.y, frame.size.width, frame.size.height);
        Bitmap resultBitmap = createEmptyBitmap(maskFrame);
        Canvas resultCanvas = new Canvas(resultBitmap);
        Paint mixPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        resultCanvas.drawBitmap(createRadiusMask(maskFrame, cornerRadius), 0, 0, mixPaint);
        mixPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        if (bitmapColor != null) {
            mixPaint.setColorFilter(new PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN));
        }
        CALayerBitmapPainter.drawBitmap(resultCanvas, maskFrame, bitmap, bitmapGravity, mixPaint);
        sPaint.reset();
        if (bitmapColor != null) {
            sPaint.setColorFilter(new PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN));
        }
        canvas.drawBitmap(resultBitmap, 0, 0, sPaint);
    }

    private static void drawRoundRectBorder(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);
        float borderWidth = (float) layer.getBorderWidth() * scaledDensity;
        float cornerRadius = (float) layer.getCornerRadius() * scaledDensity;
        float halfBorderW = borderWidth / 2.0f;

        sPaint.reset();
        sPaint.setAntiAlias(true);
        sPaint.setStyle(Paint.Style.STROKE);
        sPaint.setStrokeWidth(borderWidth);
        sPaint.setColor(layer.getBorderColor().toInt());
        RectF rectFCopyed = frame.shrinkToRectF(halfBorderW, origin);
        if (layer.getShadowRadius() > 0){
            float shadowX = (float) layer.getShadowX() * scaledDensity;
            float shadowY = (float) layer.getShadowY() * scaledDensity;
            rectFCopyed = new RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY);
        }
        canvas.drawRoundRect(rectFCopyed, cornerRadius, cornerRadius, sPaint);
    }

    private static void drawRect(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);


        sPaint.reset();
        sPaint.setColor(layer.getBackgroundColor().toInt());
        RectF rectFCopyed = frame.toRectF(origin);
        if (layer.getShadowRadius() > 0){
            float shadowRadius = (float) layer.getShadowRadius() * scaledDensity;
            float shadowX = (float) layer.getShadowX() * scaledDensity;
            float shadowY = (float) layer.getShadowY() * scaledDensity;
            sPaint.setShadowLayer(shadowRadius, shadowX, shadowY, layer.getShadowColor().toInt());
            rectFCopyed = new RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY);
        }
        canvas.drawRect(rectFCopyed, sPaint);
    }

    private static void drawRectBitmap(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        Bitmap bitmap = layer.getBitmap();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);
        UIColor bitmapColor = layer.getBitmapColor();

        sPaint.reset();
        if (bitmapColor != null) {
            sPaint.setColorFilter(new PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN));
        }
        CGRect bitmapFrame = new CGRect(origin.x, origin.y, frame.size.width, frame.size.height);
        CALayerBitmapPainter.drawBitmap(canvas, bitmapFrame, bitmap, layer.getBitmapGravity(), sPaint);
        sPaint.setColorFilter(null);
    }

    private static void drawRectBorder(Canvas canvas, CALayer layer){
        CGRect frameRaw = layer.getFrame();
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        CGRect frame = new CGRect(frameRaw.getX() * scaledDensity, frameRaw.getY() * scaledDensity, frameRaw.getWidth() * scaledDensity, frameRaw.getHeight() * scaledDensity);
        float borderWidth = (float) layer.getBorderWidth() * scaledDensity;
        float halfBorderW = borderWidth / 2.0f;

        sPaint.reset();
        sPaint.setStyle(Paint.Style.STROKE);
        sPaint.setStrokeWidth(borderWidth);
        sPaint.setColor(layer.getBorderColor().toInt());
        RectF rectFCopyed = frame.shrinkToRectF(halfBorderW, origin);
        if (layer.getShadowRadius() > 0){
            float shadowX = (float) layer.getShadowX() * scaledDensity;
            float shadowY = (float) layer.getShadowY() * scaledDensity;
            rectFCopyed = new RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY);
        }
        canvas.drawRect(rectFCopyed, sPaint);
    }

    private static Bitmap createEmptyBitmap(CGRect rect){
        int bitmapW = (int) (rect.size.width + rect.origin.x);
        int bitmapH = (int) (rect.size.height + rect.origin.y);
        Bitmap bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888);
        return bitmap;
    }

    private static Bitmap createEmptyBitmap(CALayer layer){
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        int bitmapW = (int) (layer.getFrame().size.width * scaledDensity  + origin.x);
        int bitmapH = (int) (layer.getFrame().size.height * scaledDensity + origin.y);
        Bitmap bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888);
        return bitmap;
    }

    private static Bitmap createBitmapWithLayerTree(CALayer layer){
        Bitmap bitmap = createEmptyBitmap(layer);
        layer.drawLayerTreeInCanvas(new Canvas(bitmap));
        return bitmap;
    }

    private static Bitmap createBitmapWithMaskLayer(CALayer layer){
        CGPoint origin = CALayer.calcOriginInSuperCoordinate(layer);
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        float cornerRaidus = (float) layer.getCornerRadius() * scaledDensity;
        int bitmapW = (int) (layer.getFrame().size.width * scaledDensity  + origin.x);
        int bitmapH = (int) (layer.getFrame().size.height * scaledDensity + origin.y);
        Bitmap bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888);
        RectF rectF = new RectF((float)origin.x, (float) origin.y, bitmapW, bitmapH);
        new Canvas(bitmap).drawRoundRect(rectF, cornerRaidus, cornerRaidus, new Paint(Paint.ANTI_ALIAS_FLAG));
        return bitmap;
    }

    private static Bitmap createBitmapThatApplyMask(CALayer layer, Bitmap srcBitmap, Bitmap maskBitmap){
        sPaint.reset();
        Bitmap bitmapMixed = createEmptyBitmap(layer);
        Canvas canvasMixed = new Canvas(bitmapMixed);
        if (layer.getTransforms() != null && layer.getTransforms().length > 0){
            Matrix matrix = createMatrix(layer);
            if (layer.getClipToBounds()){
                canvasMixed.drawBitmap(maskBitmap, matrix, sPaint);
                sPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
            }
            canvasMixed.drawBitmap(srcBitmap, matrix, sPaint);
            sPaint.setXfermode(null);
        }
        else {
            canvasMixed.drawBitmap(maskBitmap, 0, 0, sPaint);
            sPaint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
            canvasMixed.drawBitmap(srcBitmap, 0, 0, sPaint);
            sPaint.setXfermode(null);
        }
        return bitmapMixed;
    }

    private static Bitmap createRadiusMask(@NonNull CGRect rect, double radius){
        Bitmap maskBitmap = Bitmap.createBitmap((int)(rect.size.width+rect.origin.x), (int)(rect.size.height+rect.origin.y), Bitmap.Config.ARGB_8888);
        Canvas maskCanvas = new Canvas(maskBitmap);
        Paint maskPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        maskCanvas.drawRoundRect(rect.toRectF(), (float) radius, (float) radius, maskPaint);
        return maskBitmap;
    }

    private static Matrix createMatrix(CALayer layer){
        float scaledDensity = (float) UIScreen.mainScreen.scale();
        CGTransform[] transforms = layer.getTransforms();
        Matrix matrix = new Matrix();
        if (transforms == null || transforms.length == 0){
            return matrix;
        }
        RectF rectF = layer.getFrame().toRectF();
        for (CGTransform transform : transforms){
            if (!transform.enable){
                continue;
            }
            if (transform instanceof CGTransformRotation){
                matrix.preRotate((float) ((CGTransformRotation) transform).angle, rectF.centerX() * (float) scaledDensity, rectF.centerY() * (float) scaledDensity);
            }
            else if (transform instanceof CGTransformTranslation){
                CGTransformTranslation translation = (CGTransformTranslation) transform;
                matrix.postTranslate((float) translation.tx, (float) translation.ty);
            }
            else if (transform instanceof CGTransformScale){
                CGTransformScale scale = (CGTransformScale)transform;
                matrix.postScale((float) scale.sx, (float) scale.sy, rectF.centerX() * (float) scaledDensity, rectF.centerY() * (float) scaledDensity);
            }
            else if (transform instanceof CGTransformMatrix){
                // @TODO
            }
        }
        return matrix;
    }

}

File: library/src/main/java/com/yy/codex/coreanimation/CALayerPainter.kt
package com.yy.codex.coreanimation
import android.graphics.*
import com.yy.codex.uikit.*

/**
 * Created by adi on 17/1/17.
 */

internal object CALayerPainter {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)

    fun drawLayerTree(layer: CALayer, canvas: Canvas) {
        val srcBitmap = createBitmapWithLayerTree(layer)
        val maskBitmap = createBitmapWithMaskLayer(layer)
        val resultBitmap = createBitmapThatApplyMask(layer, srcBitmap, maskBitmap)
        paint.reset()
        canvas.drawBitmap(resultBitmap, 0f, 0f, paint)
    }

    fun drawCurrentLayer(layer: CALayer, canvas: Canvas) {
        if (layer.hidden) {
            return
        }
        if (layer.cornerRadius > 0) {
            drawRoundRect(canvas, layer)
            if (layer.bitmap != null) {
                drawRoundRectBitmap(canvas, layer)
            }
            if (layer.borderWidth > 0) {
                drawRoundRectBorder(canvas, layer)
            }
        } else {
            drawRect(canvas, layer)
            if (layer.bitmap != null) {
                drawRectBitmap(canvas, layer)
            }
            if (layer.borderWidth > 0) {
                drawRectBorder(canvas, layer)
            }
        }

        if (layer.mask != null) {
            // @TODO
        }
    }

    /* support method */

    private fun drawRoundRect(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)
        val borderWidth = layer.borderWidth.toFloat() * scaledDensity
        val cornerRadius = layer.cornerRadius.toFloat() * scaledDensity
        val halfBorderW = borderWidth / 2.0f

        paint.reset()
        paint.isAntiAlias = true
        paint.color = layer.backgroundColor.toInt()
        var rectFCopyed = frame.shrinkToRectF(halfBorderW, origin)
        if (layer.shadowRadius > 0) {
            val shadowRadius = layer.shadowRadius.toFloat() * scaledDensity
            val shadowX = layer.shadowX.toFloat() * scaledDensity
            val shadowY = layer.shadowY.toFloat() * scaledDensity
            paint.setShadowLayer(shadowRadius, shadowX, shadowY, layer.shadowColor.toInt())
            rectFCopyed = RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY)
        }
        canvas.drawRoundRect(rectFCopyed, cornerRadius, cornerRadius, paint)
    }

    private fun drawRoundRectBitmap(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val bitmap = layer.bitmap
        val bitmapGravity = layer.bitmapGravity
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)
        val cornerRadius = layer.cornerRadius.toFloat() * scaledDensity
        val bitmapColor = layer.bitmapColor

        val maskFrame = CGRect(origin.x, origin.y, frame.size.width, frame.size.height)
        val resultBitmap = createEmptyBitmap(maskFrame)
        val resultCanvas = Canvas(resultBitmap)
        val mixPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        resultCanvas.drawBitmap(createRadiusMask(maskFrame, cornerRadius.toDouble()), 0f, 0f, mixPaint)
        mixPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
        if (bitmapColor != null) {
            mixPaint.colorFilter = PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN)
        }
        bitmap?.let {
            CALayerBitmapPainter.drawBitmap(resultCanvas, maskFrame, it, bitmapGravity, mixPaint)
        }
        paint.reset()
        if (bitmapColor != null) {
            paint.colorFilter = PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN)
        }
        canvas.drawBitmap(resultBitmap, 0f, 0f, paint)
    }

    private fun drawRoundRectBorder(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)
        val borderWidth = layer.borderWidth.toFloat() * scaledDensity
        val cornerRadius = layer.cornerRadius.toFloat() * scaledDensity
        val halfBorderW = borderWidth / 2.0f

        paint.reset()
        paint.isAntiAlias = true
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = borderWidth
        paint.color = layer.borderColor.toInt()
        var rectFCopyed = frame.shrinkToRectF(halfBorderW, origin)
        if (layer.shadowRadius > 0) {
            val shadowX = layer.shadowX.toFloat() * scaledDensity
            val shadowY = layer.shadowY.toFloat() * scaledDensity
            rectFCopyed = RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY)
        }
        canvas.drawRoundRect(rectFCopyed, cornerRadius, cornerRadius, paint)
    }

    private fun drawRect(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)


        paint.reset()
        paint.color = layer.backgroundColor.toInt()
        var rectFCopyed = frame.toRectF(origin)
        if (layer.shadowRadius > 0) {
            val shadowRadius = layer.shadowRadius.toFloat() * scaledDensity
            val shadowX = layer.shadowX.toFloat() * scaledDensity
            val shadowY = layer.shadowY.toFloat() * scaledDensity
            paint.setShadowLayer(shadowRadius, shadowX, shadowY, layer.shadowColor.toInt())
            rectFCopyed = RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY)
        }
        canvas.drawRect(rectFCopyed, paint)
    }

    private fun drawRectBitmap(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val bitmap = layer.bitmap
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)
        val bitmapColor = layer.bitmapColor
        paint.reset()
        if (bitmapColor != null) {
            paint.colorFilter = PorterDuffColorFilter(bitmapColor.toInt(), PorterDuff.Mode.SRC_IN)
        }
        val bitmapFrame = CGRect(origin.x, origin.y, frame.size.width, frame.size.height)
        bitmap?.let {
            CALayerBitmapPainter.drawBitmap(canvas, bitmapFrame, it, layer.bitmapGravity, paint)
        }
        paint.colorFilter = null
    }

    private fun drawRectBorder(canvas: Canvas, layer: CALayer) {
        val frameRaw = layer.frame
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val frame = CGRect(frameRaw.x * scaledDensity, frameRaw.y * scaledDensity, frameRaw.width * scaledDensity, frameRaw.height * scaledDensity)
        val borderWidth = layer.borderWidth.toFloat() * scaledDensity
        val halfBorderW = borderWidth / 2.0f

        paint.reset()
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = borderWidth
        paint.color = layer.borderColor.toInt()
        var rectFCopyed = frame.shrinkToRectF(halfBorderW, origin)
        if (layer.shadowRadius > 0) {
            val shadowX = layer.shadowX.toFloat() * scaledDensity
            val shadowY = layer.shadowY.toFloat() * scaledDensity
            rectFCopyed = RectF(rectFCopyed.left, rectFCopyed.top, rectFCopyed.right - shadowX, rectFCopyed.bottom - shadowY)
        }
        canvas.drawRect(rectFCopyed, paint)
    }

    private fun createEmptyBitmap(rect: CGRect): Bitmap {
        val bitmapW = (rect.size.width + rect.origin.x).toInt()
        val bitmapH = (rect.size.height + rect.origin.y).toInt()
        val bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888)
        return bitmap
    }

    private fun createEmptyBitmap(layer: CALayer): Bitmap {
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val bitmapW = (layer.frame.size.width * scaledDensity + origin.x).toInt()
        val bitmapH = (layer.frame.size.height * scaledDensity + origin.y).toInt()
        val bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888)
        return bitmap
    }

    private fun createBitmapWithLayerTree(layer: CALayer): Bitmap {
        val bitmap = createEmptyBitmap(layer)
        layer.drawLayerTreeInCanvas(Canvas(bitmap))
        return bitmap
    }

    private fun createBitmapWithMaskLayer(layer: CALayer): Bitmap {
        val origin = CALayer.calcOriginInSuperCoordinate(layer)
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val cornerRaidus = layer.cornerRadius.toFloat() * scaledDensity
        val bitmapW = (layer.frame.size.width * scaledDensity + origin.x).toInt()
        val bitmapH = (layer.frame.size.height * scaledDensity + origin.y).toInt()
        val bitmap = Bitmap.createBitmap(bitmapW, bitmapH, Bitmap.Config.ARGB_8888)
        val rectF = RectF(origin.x.toFloat(), origin.y.toFloat(), bitmapW.toFloat(), bitmapH.toFloat())
        Canvas(bitmap).drawRoundRect(rectF, cornerRaidus, cornerRaidus, Paint(Paint.ANTI_ALIAS_FLAG))
        return bitmap
    }

    private fun createBitmapThatApplyMask(layer: CALayer, srcBitmap: Bitmap, maskBitmap: Bitmap): Bitmap {
        paint.reset()
        val bitmapMixed = createEmptyBitmap(layer)
        val canvasMixed = Canvas(bitmapMixed)
        if (layer.transforms.size > 0) {
            val matrix = createMatrix(layer)
            if (layer.clipToBounds) {
                canvasMixed.drawBitmap(maskBitmap, matrix, paint)
                paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
            }
            canvasMixed.drawBitmap(srcBitmap, matrix, paint)
            paint.xfermode = null
        } else {
            canvasMixed.drawBitmap(maskBitmap, 0f, 0f, paint)
            paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)
            canvasMixed.drawBitmap(srcBitmap, 0f, 0f, paint)
            paint.xfermode = null
        }
        return bitmapMixed
    }

    private fun createRadiusMask(rect: CGRect, radius: Double): Bitmap {
        val maskBitmap = Bitmap.createBitmap((rect.size.width + rect.origin.x).toInt(), (rect.size.height + rect.origin.y).toInt(), Bitmap.Config.ARGB_8888)
        val maskCanvas = Canvas(maskBitmap)
        val maskPaint = Paint(Paint.ANTI_ALIAS_FLAG)
        maskCanvas.drawRoundRect(rect.toRectF(), radius.toFloat(), radius.toFloat(), maskPaint)
        return maskBitmap
    }

    private fun createMatrix(layer: CALayer): Matrix {
        val scaledDensity = UIScreen.mainScreen.scale().toFloat()
        val transforms = layer.transforms
        val matrix = Matrix()
        if (transforms == null || transforms.size == 0) {
            return matrix
        }
        val rectF = layer.frame.toRectF()
        transforms
                .filter { it.enable }
                .forEach {
                    if (it is CGTransformRotation) {
                        matrix.preRotate(it.angle.toFloat(), rectF.centerX() * scaledDensity, rectF.centerY() * scaledDensity)
                    } else if (it is CGTransformTranslation) {
                        matrix.postTranslate(it.tx.toFloat(), it.ty.toFloat())
                    } else if (it is CGTransformScale) {
                        matrix.postScale(it.sx.toFloat(), it.sy.toFloat(), rectF.centerX() * scaledDensity, rectF.centerY() * scaledDensity)
                    } else if (it is CGTransformMatrix) {
                        // @TODO
                    }
                }
        return matrix
    }

}
